File: Swift-C++-Testsuite/test/structsAndClasses/Swift/Classes/FailableInit.swift
@main
struct MyMain {

    static func main() {
        let someCreature = Animal(species: "Giraffe")
        // someCreature is of type Animal?, not Animal

        if let giraffe = someCreature {
            print("An animal was initialized with a species of \(giraffe.species)")
        }

        let anonymousCreature = Animal(species: "")
        // anonymousCreature is of type Animal?, not Animal

        if anonymousCreature == nil {
            print("The anonymous creature couldn't be initialized")
        }
    }
}

struct Animal {
    let species: String
    init?(species: String) {
        if species.isEmpty { return nil }
        self.species = species
    }
}

File: Swift-C++-Testsuite/test/structsAndClasses/C++/Classes/FailableInit.cpp


#include <iostream>

struct Animal
{
    public: 
        std::string species;
        static Animal* create(std::string species) {
        if (species.empty())
            {
                return nullptr;
            }
            return new Animal(species);
        }

    private:

    Animal(std::string species)
    {

        this->species = species;
    }
};

int main()
{
    Animal *someCreature = Animal::create("Giraffe");
    if (someCreature)
    {
        std::cout << "An animal was initialized with a species of " << someCreature->species << std::endl;
    }

    Animal *anonymousCreature = Animal::create("");
    if (!anonymousCreature)
    {
        std::cout << "The anonymous creature couldn't be initialized" << std::endl;
    }

    delete someCreature;
    delete anonymousCreature;

    return 0;
}